airbyte-scheduler/scheduler-persistence/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java

```diff
@@ -340,17 +341,70 @@ private Optional<Job> getJobOptional(final DSLContext ctx, final long jobId) {
     return getJobFromResult(ctx.fetch(BASE_JOB_SELECT_AND_JOIN + "WHERE jobs.id = ?", jobId));
   }
 
+  @Override
+  public Long getJobCount(final Set<ConfigType> configTypes, final String connectionId) throws IOException {
+    return jobDatabase.query(ctx -> ctx.selectCount().from(JOBS)
+        .where(JOBS.CONFIG_TYPE.cast(String.class).in(Sqls.toSqlNames(configTypes)))
+        .and(JOBS.SCOPE.eq(connectionId))
+        .fetchOne().into(Long.class));
+  }
+
   @Override
   public List<Job> listJobs(final ConfigType configType, final String configId, final int pagesize, final int offset) throws IOException {
     return listJobs(Set.of(configType), configId, pagesize, offset);
   }
 
   @Override
   public List<Job> listJobs(final Set<ConfigType> configTypes, final String configId, final int pagesize, final int offset) throws IOException {
-    final String jobsSubquery = "(SELECT * FROM jobs WHERE CAST(jobs.config_type AS VARCHAR) in " + Sqls.toSqlInFragment(configTypes)
-        + " AND jobs.scope = '" + configId + "' ORDER BY jobs.created_at DESC, jobs.id DESC LIMIT " + pagesize + " OFFSET " + offset + ") AS jobs";
-    return jobDatabase.query(ctx -> getJobsFromResult(ctx.fetch(
-        jobSelectAndJoin(jobsSubquery) + ORDER_BY_JOB_TIME_ATTEMPT_TIME)));
+    return jobDatabase.query(ctx -> {
+      final String jobsSubquery = "(" + ctx.select(DSL.asterisk()).from(JOBS)
+          .where(JOBS.CONFIG_TYPE.cast(String.class).in(DSL.inline(Sqls.toSqlInFragmentInline(configTypes))))
+          .and(JOBS.SCOPE.eq(DSL.inline(configId)))
+          .orderBy(JOBS.CREATED_AT.desc(), JOBS.ID.desc())
+          .limit(DSL.inline(pagesize))
+          .offset(DSL.inline(offset))
+          .getSQL() + ") AS jobs";
+
+      return getJobsFromResult(ctx.fetch(jobSelectAndJoin(jobsSubquery) + ORDER_BY_JOB_TIME_ATTEMPT_TIME));
+    });
+  }
+
+  @Override
+  public List<Job> listJobsStartingWithId(final Set<ConfigType> configTypes, final String connectionId, final long startingJobId, final int pagesize)
+      throws IOException {
+    final Optional<OffsetDateTime> startingJobCreatedAt = jobDatabase.query(ctx -> ctx.select(JOBS.CREATED_AT).from(JOBS)
+        .where(JOBS.CONFIG_TYPE.cast(String.class).in(Sqls.toSqlNames(configTypes)))
+        .and(JOBS.SCOPE.eq(connectionId))
+        .and(JOBS.ID.eq(startingJobId))
+        .stream()
+        .findFirst()
+        .map(record -> record.get(JOBS.CREATED_AT, OffsetDateTime.class)));
+
+    if (startingJobCreatedAt.isEmpty()) {
+      return List.of();
+    }
+
+    final int countIncludingStartingJob = jobDatabase.query(ctx -> ctx.selectCount().from(JOBS)
+        .where(JOBS.CONFIG_TYPE.cast(String.class).in(Sqls.toSqlNames(configTypes)))
+        .and(JOBS.SCOPE.eq(connectionId))
+        .and(JOBS.CREATED_AT.greaterOrEqual(startingJobCreatedAt.get()))
+        .fetchOne().into(int.class));
+
+    // always want to return at least `pagesize` number of jobs
```

Comment:
I don't understand this part, why are we going to fill more jobs unrelated to jobId? I'd expect that if we do not have enough jobs to fill the page, it's fine. Otherwise we are mixing two different sets of jobs which sounds more confusing in the end.

Reply:
This is specific for the frontend use case, since we are trying to avoid the situation where if someone links something like the second or third most recent job, then only 2 or 3 jobs would be loaded on the page. This could be confusing if there are many more jobs in that connection, since we use a page size of 25. So having the UI only load a few jobs when there are actually many more would be strange

Reply:
Since the `listJobs()` call always returns jobs from the most recent in descending order of creation time, and we are writing the `countIncludingStartingJob` query in a similar way, then we should be guaranteed that if that count is less then the page size, then the `listJobs()` call will include the startingJob that we want